net: coap: add support for CoAP over reliable transports

mpenate-ellenbytech opened the pull request:
Add implementation of RFC 8323 (CoAP over TCP). This includes:
- TCP-specific packet parsing and creation functions
- BERT (Block-wise Extension for Reliable Transport) support
- CSM signaling code (7.01) and options
- New Kconfig options for reliable transport configuration

rlubos reviewed (commented):
An initial pass on the PR.

For the final version, we should also aim to have some sample covering this functionality (or ideally some tests if possible), and documentation.

rlubos commented on subsys/net/lib/coap/coap_client_tcp.c:
```diff
@@ -0,0 +1,1068 @@
+	}
+
+	/* Step 6: Complete packet received, reset offset for next packet */
+	client->recv_offset = 0;
```
Could it be the case that we receive two messages from the server one after another?
Note that now `zsock_recv()` can read up to to the buffer size number of bytes, so theoretically we could've read more than a single message in a single `zsock_recv()` call. If we reset the offset here, we might lose the other message.
If `total_expected` is the size of the currently processed message, we should check if `offset > total_expected` and `memmove()` the remaining bytes to the beginning of the buffer. And instead of clearing the offset, decrease it by `total_expected` value.

mpenate-ellenbytech replied:
See comment above, removed setting `recv_offset` to 0 here and instead after handling the response in the polling function I adjust the value there.

mpenate-ellenbytech commented:
I added rudimentary handling of the CSM max message size option but the RFC specifies some, lets say complicated implementation requirements that don't fit in with the original idea of CoAP client so I pushed the basic handling and come here to continue the discussion.

[Section 5.3.2](https://datatracker.ietf.org/doc/html/rfc8323#section-5.3.2) on the enabling of block-wise transfers via CSMs:
> If a Max-Message-Size Option is indicated with a value that is
   greater than 1152 (in the same CSM or a different CSM), the
   Block-Wise-Transfer Option also indicates support for BERT (see
   [Section 6](https://datatracker.ietf.org/doc/html/rfc8323#section-6))
And earlier in [Section 5.3](https://datatracker.ietf.org/doc/html/rfc8323#section-5.3):
> Base values are listed below for CSM options.  These are the values
   for the capability and settings before any CSMs send a modified
   value.

The base value for the [Block-Wise-Transfer Capability Option](https://datatracker.ietf.org/doc/html/rfc8323#section-5.3.2) is unspecified meaning that generic block-wise transfer and BERT block-wise transfer should technically be runtime-disabled until this CSM is negotiated with the server correct?

If we were to do this, then the "context" of the socket is now more entangled with CoAP client than it was before. When using a new socket fd with the coap client, the developer needs to ask themselves "Is this socket context already negotiated with basic CSMs?" and inform the client of these previous negotiations if they exist which IMO isn't the best.

@rlubos what do you think about integrating more of the socket connection context into the management handled by CoAP client? Personally I could see a world where a client is initialized with a basic int-returning callback that should return a `zsock_connect`-ed socket FD. Then if anything ever goes wrong with the socket, the client can close it and request a new one be opened using the callback? Then upon using the callback it can automatically handle the CSMs and the book-keeping needed for it. This would restrict a client to one connected context at a time but that restriction already exists in the form of a single socket fd in the UDP client.

rlubos commented on subsys/net/lib/coap/coap_client_tcp.c:
```diff
@@ -0,0 +1,1078 @@
+				client->recv_offset -= pkt_len;
+			} else {
+				client->recv_offset = 0;
+			}
```
I don't think there's a need, the RX buffer will only be updated from the CoAP TCP client thread so there's likely no need for extra mutex protection. That client mutex to my understanding protects the internal request contexts, where those could be updated from different threads (like for example when cancelling request).

rlubos commented:
> The base value for the [Block-Wise-Transfer Capability Option](https://datatracker.ietf.org/doc/html/rfc8323#section-5.3.2) is unspecified meaning that generic block-wise transfer and BERT block-wise transfer should technically be runtime-disabled until this CSM is negotiated with the server correct?

That's my understanding as well - lack of Block-wise-transfer option in CSM indicates no block transfer support by the sender, so block transfers should not be used over such connection. Probably good-enough approach for initial implementation.

> @rlubos what do you think about integrating more of the socket connection context into the management handled by CoAP client? Personally I could see a world where a client is initialized with a basic int-returning callback that should return a zsock_connect-ed socket FD. Then if anything ever goes wrong with the socket, the client can close it and request a new one be opened using the callback? 

I agree, client context in CoAP over TCP is practically bound to the underlying TCP/TLS connection, so it makes sense to integrate the client with the socket more closely. I'd imagine we'd have two additional session-related functions for CoAP over TCP, like:
* `coap_client_tcp_connect()`
* `coap_client_tcp_close()`

`coap_client_tcp_connect()` would make a connection at the socket level, and initialize CoAP session (i.e. send CSM). `coap_client_tcp_close()` would close the session (I see in the RFC that there are also various signals to notify this, instead of just closing the connection at the transport level).

As you say, if any error is returned when making requests or via response callback, the session should be considered invalid and restarted by the app.

Socket creation could be handled via callback as you suggests, or simply a FD to a created (not yet connected) socket could be passed to `coap_client_tcp_connect()`. Either should work, and I agree that this should be easier than having to cover socket creation with all the details within the library.

> Then upon using the callback it can automatically handle the CSMs and the book-keeping needed for it. This would restrict a client to one connected context at a time but that restriction already exists in the form of a single socket fd in the UDP client.

I think it's quite natural that single client context maintains a single connection. If needed, multiple client contexts can be created (at least that's how it works with regular CoAP client).

mpenate-ellenbytech commented:
> > The base value for the [Block-Wise-Transfer Capability Option](https://datatracker.ietf.org/doc/html/rfc8323#section-5.3.2) is unspecified meaning that generic block-wise transfer and BERT block-wise transfer should technically be runtime-disabled until this CSM is negotiated with the server correct?
> 
> That's my understanding as well - lack of Block-wise-transfer option in CSM indicates no block transfer support by the sender, so block transfers should not be used over such connection. Probably good-enough approach for initial implementation.

Added a bool to the client struct `blockwise_enabled` to track status from CSMs received. Added a connect and close function. Trying to send a request when there is no socket (ie. the client needs to connect)  returns `-ENOTCONN`. Poll errors result in the connection being closed and the socket reset to `-1` (causing above mentioned `-ENOTCONN`). 

> I agree, client context in CoAP over TCP is practically bound to the underlying TCP/TLS connection, so it makes sense to integrate the client with the socket more closely. I'd imagine we'd have two additional session-related functions for CoAP over TCP...

The connect function calls a simple callback currently (callback stored in the client struct)
```
	coap_client_tcp_socket_config_cb_t socket_config_cb;
	void *socket_config_cb_user_data;
```

In this callback the application developer can call `zsock_setsockopt` to configure the socket fd (ie. binding to an interface, specifying TLS credential tags, configuring keep alive, etc.). If this callback returns a negative value that value is returned as the result from `coap_client_tcp_connect`. Otherwise after the callback, zsock_connect is attempted (**currently** within the thread context of an application thread, not CoAP thread, should this change?). Failures handled still (closing/cleaning the socket fd), otherwise on successful connect a CSM is immediately queued as the spec states the client should always send a CSM message.

Some public APIs are missing documentation.

I left the CSM function declaration in the header as the runtime application should technically be able to change the MMS at runtime and throughout an active connection.

> I see in the RFC that there are also various signals to notify this, instead of just closing the connection at the transport level

Looking at targeting some of these next.

rlubos commented:
> Otherwise after the callback, zsock_connect is attempted (currently within the thread context of an application thread, not CoAP thread, should this change?)

I don't think that's a problem, probably even better as this allows to detect connect issues early and return an error from `coap_client_tcp_connect()`.

rlubos commented on include/zephyr/net/coap_client_tcp.h:
```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
```
Please update the copyright and date

rlubos commented on include/zephyr/net/coap_client_tcp.h:
```diff
@@ -0,0 +1,234 @@
+/**
+ * @brief Representation for CoAP TCP client response data.
+ */
+struct coap_client_tcp_response_data {
```
Any thoughts to this? It still seems to me that part of the concepts (request/response representation, option etc.) could be shared between implementations and only differentiate at the API level.

rlubos commented on subsys/net/lib/coap/coap.c:
```diff
@@ -2136,3 +2171,433 @@ void coap_set_transmission_parameters(const struct coap_transmission_parameters
+		return -EINVAL;
+	}
+
+	if (options) {
```
`options != NULL`, please also fix similar cases elsewhere.

rlubos commented on subsys/net/lib/coap/coap_client_tcp.c:
```diff
@@ -0,0 +1,1362 @@
+static int coap_client_tcp_init_request(struct coap_client_tcp *client,
+					struct coap_client_tcp_request *req,
+					struct coap_client_tcp_internal_request *internal_req,
+					bool reconstruct)
```
Do we still need the `reconstruct` flag, given each request has a separate buffer. Note this was removed from the regular coap_client.

rlubos commented on subsys/net/lib/coap/coap_client_tcp.c:
```diff
@@ -0,0 +1,1362 @@
+	struct coap_client_tcp_internal_request *internal_req;
+	struct coap_option csm_option;
+	uint8_t cms_option_num;
+
```
nit: unneeded empty line.

rlubos reviewed (commented):
> @rlubos I think this is at the point that it could be promoted from draft PR. LMK your thoughts.

Yeah I think it can be moved out of draft already.

Generally the implementation looks pretty good, although I still wanted to discuss a bit whether we could share some common data representations between regular coap_client and TCP variant. But if there are good reasons not to, no strong push from my side.

Also I think we need basic example of the new library usage (separate sample most likely). A doc update would also be appreciated.

rlubos reviewed (commented):
Nice, thanks for the sample update, I think it's much easier to follow now w/o the built-in server overhead. One final suggestion, but otherwise LGTM. You could possible pull it out of draft already?

mpenate-ellenbytech committed:
net: coap: add support for CoAP over reliable transports

Add CoAP over TCP/TLS framing, parsing, and client support as
specified in RFC 8323. This includes:

- TCP message framing with variable-length header encoding
- CSM (Capabilities and Settings Message) exchange
- Signaling codes: Ping/Pong, Release, Abort
- CoAP TCP client with blockwise transfer support
- New Kconfig options under CONFIG_COAP_OVER_RELIABLE_TRANSPORT
- Documentation update for CoAP client API

Signed-off-by: Marcus Penate <marcus.penate@ellenbytech.com>

mpenate-ellenbytech committed:
samples: net: coap: add CoAP over TCP client sample

Add a sample demonstrating the CoAP over TCP client API (RFC 8323).
The sample connects to an external CoAP TCP server (e.g. aiocoap)
and exercises CSM exchange, Ping/Pong, GET requests, and graceful
session release.

Signed-off-by: Marcus Penate <marcus.penate@ellenbytech.com>

rlubos reviewed (dismissed):
LGTM, Thanks

jukkar commented on subsys/net/lib/coap/coap.c:
```diff
@@ -2135,3 +2170,433 @@ void coap_set_transmission_parameters(const struct coap_transmission_parameters
+		offset_to_code += (1 << (header_ext_len - COAP_TCP_HEADER_LEN_EXT_1B));
+	}
+
+	uint8_t code = cpkt->data[offset_to_code];
```
ditto

jukkar commented on subsys/net/lib/coap/coap_client_tcp.c:
```diff
@@ -0,0 +1,1364 @@
+		}
+
+		/* Parse extended length */
+		uint32_t ext_val = 0;
```
Declaration at the start of the block please.

jukkar reviewed (commented):
Some style issues comments.

The bigger changes could be done in the subsequent PRs if you are willing to commit to develop the code further. Like
* investigating to get rid of handler thread and replace it by socket service api
* investigating to send events in various stages in the connection so that we can avoid having hard coded sleeps in the application

mpenate-ellenbytech commented on include/zephyr/net/coap_client_tcp.h:
```diff
@@ -0,0 +1,363 @@
+	/** Option len */
+	uint8_t len;
+	/** Buffer for the length */
+	uint8_t value[12];
```
Want me to add one to the general coap header and replace the constant 12 in the original coap client header as well?

jukkar replied:
What ever makes sense, just trying to avoid magic constants.

rlubos reviewed (dismissed):
Compliance failure needs fixing:
>  LINE_SPACING: Missing a blank line after declarations
File:subsys/net/lib/coap/coap.c
Line:2400
